File: python_backend/services/pricing_service.py
"""
Pricing Service - Utility pricing and cost optimization
"""
from datetime import datetime, time, timedelta
from typing import Dict, List, Any, Optional
from dataclasses import dataclass
from enum import Enum
import sys
import os

sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

from config import settings


class TimePeriod(Enum):
    PEAK = "peak"
    MID_PEAK = "mid_peak"
    OFF_PEAK = "off_peak"


@dataclass
class TimeWindow:
    start_hour: int
    end_hour: int
    avg_price: float
    
    def to_dict(self) -> Dict[str, Any]:
        return {
            "start_hour": self.start_hour,
            "end_hour": self.end_hour,
            "avg_price": round(self.avg_price, 4),
            "time_range": f"{self.start_hour:02d}:00 - {self.end_hour:02d}:00"
        }


class PricingService:
    """
    Utility Pricing Service for electricity cost optimization.
    Supports time-of-use pricing and real-time pricing integration.
    """
    
    def __init__(self):
        self.base_price_per_kwh = settings.UTILITY_PRICE_PER_KWH
        self.currency = settings.UTILITY_CURRENCY
        
        # Time-of-use pricing tiers
        self.time_of_use_prices = {
            TimePeriod.PEAK: self.base_price_per_kwh * 1.5,      # $0.18
            TimePeriod.MID_PEAK: self.base_price_per_kwh * 1.2,  # $0.144
            TimePeriod.OFF_PEAK: self.base_price_per_kwh * 0.7   # $0.084
        }
    
    def get_current_price_per_kwh(self) -> float:
        """Get current price per kWh based on time-of-use"""
        now = datetime.now()
        return self.get_price_for_time(now.time(), now.weekday())
    
    def get_price_for_time(self, t: time, day_of_week: int) -> float:
        """Get price for a specific time"""
        period = self._get_time_period(t, day_of_week)
        return self.time_of_use_prices[period]
    
    def _get_time_period(self, t: time, day_of_week: int) -> TimePeriod:
        """Determine time period based on time and day"""
        is_weekend = day_of_week >= 5  # Saturday=5, Sunday=6
        
        if is_weekend:
            if t.hour < 6 or t.hour >= 22:
                return TimePeriod.OFF_PEAK
            return TimePeriod.MID_PEAK
        
        # Weekday pricing
        if t.hour < 6 or t.hour >= 21:
            return TimePeriod.OFF_PEAK
        
        if t.hour < 10 or t.hour >= 17:
            return TimePeriod.MID_PEAK
        
        return TimePeriod.PEAK
    
    def calculate_running_cost(self, wattage: float, duration_hours: float) -> float:
        """Calculate cost of current consumption"""
        kwh = wattage / 1000 * duration_hours
        return kwh * self.get_current_price_per_kwh()
    
    def calculate_daily_cost(self, average_wattage: float) -> float:
        """Calculate daily cost estimate"""
        total_cost = 0
        
        for hour in range(24):
            price = self.get_price_for_time(time(hour, 0), datetime.now().weekday())
            hourly_kwh = average_wattage / 1000
            total_cost += hourly_kwh * price
        
        return total_cost
    
    def calculate_monthly_cost(self, average_daily_cost: float) -> float:
        """Calculate monthly cost estimate"""
        days_in_month = datetime.now().replace(day=28) + timedelta(days=4)
        days_in_month = (days_in_month - timedelta(days=days_in_month.day)).day
        return average_daily_cost * days_in_month
    
    def get_hourly_prices_for_today(self) -> Dict[int, float]:
        """Get hourly prices for today"""
        today = datetime.now().weekday()
        return {
            hour: self.get_price_for_time(time(hour, 0), today)
            for hour in range(24)
        }
    
    def get_low_cost_windows(self) -> List[TimeWindow]:
        """Get low-cost time windows for scheduling"""
        hourly_prices = self.get_hourly_prices_for_today()
        avg_price = sum(hourly_prices.values()) / len(hourly_prices)
        low_cost_threshold = avg_price * 0.9
        
        windows = []
        current_window: Optional[TimeWindow] = None
        
        for hour in range(24):
            price = hourly_prices[hour]
            
            if price <= low_cost_threshold:
                if current_window is None:
                    current_window = TimeWindow(hour, hour + 1, price)
                else:
                    current_window = TimeWindow(
                        current_window.start_hour,
                        hour + 1,
                        (current_window.avg_price + price) / 2
                    )
            else:
                if current_window:
                    windows.append(current_window)
                    current_window = None
        
        if current_window:
            windows.append(current_window)
        
        return windows
    
    def get_high_cost_windows(self) -> List[TimeWindow]:
        """Get high-cost time windows to avoid"""
        hourly_prices = self.get_hourly_prices_for_today()
        avg_price = sum(hourly_prices.values()) / len(hourly_prices)
        high_cost_threshold = avg_price * 1.1
        
        windows = []
        current_window: Optional[TimeWindow] = None
        
        for hour in range(24):
            price = hourly_prices[hour]
            
            if price >= high_cost_threshold:
                if current_window is None:
                    current_window = TimeWindow(hour, hour + 1, price)
                else:
                    current_window = TimeWindow(
                        current_window.start_hour,
                        hour + 1,
                        (current_window.avg_price + price) / 2
                    )
            else:
                if current_window:
                    windows.append(current_window)
                    current_window = None
        
        if current_window:
            windows.append(current_window)
        
        return windows
    
    def get_optimal_run_time(self, duration_hours: int) -> Optional[TimeWindow]:
        """Get optimal time to run a high-power task"""
        low_cost_windows = self.get_low_cost_windows()
        
        best_window = None
        best_price = float('inf')
        
        for window in low_cost_windows:
            window_duration = window.end_hour - window.start_hour
            if window_duration >= duration_hours and window.avg_price < best_price:
                best_price = window.avg_price
                best_window = window
        
        return best_window
    
    def calculate_load_shifting_savings(
        self,
        wattage: float,
        duration_hours: int,
        from_hour: int,
        to_hour: int
    ) -> Dict[str, Any]:
        """Calculate potential savings from load shifting"""
        today = datetime.now().weekday()
        
        from_price = self.get_price_for_time(time(from_hour, 0), today)
        to_price = self.get_price_for_time(time(to_hour, 0), today)
        
        kwh = wattage / 1000 * duration_hours
        current_cost = kwh * from_price
        shifted_cost = kwh * to_price
        savings = current_cost - shifted_cost
        
        return {
            "current_cost": round(current_cost, 2),
            "shifted_cost": round(shifted_cost, 2),
            "savings": round(savings, 2),
            "savings_pct": round((savings / current_cost * 100) if current_cost > 0 else 0, 2),
            "currency": self.currency
        }
    
    def get_pricing_summary(self) -> Dict[str, Any]:
        """Get pricing summary and recommendations"""
        low_cost_windows = self.get_low_cost_windows()
        high_cost_windows = self.get_high_cost_windows()
        
        summary = {
            "current_price_per_kwh": round(self.get_current_price_per_kwh(), 4),
            "base_price_per_kwh": self.base_price_per_kwh,
            "currency": self.currency,
            "peak_price": self.time_of_use_prices[TimePeriod.PEAK],
            "mid_peak_price": self.time_of_use_prices[TimePeriod.MID_PEAK],
            "off_peak_price": self.time_of_use_prices[TimePeriod.OFF_PEAK],
            "low_cost_windows": [w.to_dict() for w in low_cost_windows],
            "high_cost_windows": [w.to_dict() for w in high_cost_windows]
        }
        
        if low_cost_windows:
            best_window = low_cost_windows[0]
            summary["best_time_to_consume"] = f"{best_window.start_hour:02d}:00 - {best_window.end_hour:02d}:00"
        
        return summary
    
    def get_carbon_intensity(self) -> float:
        """
        Get carbon intensity estimate (kg CO2 per kWh).
        Higher during peak hours when dirtier plants are used.
        """
        now = datetime.now()
        period = self._get_time_period(now.time(), now.weekday())
        
        carbon_rates = {
            TimePeriod.PEAK: 0.5,
            TimePeriod.MID_PEAK: 0.4,
            TimePeriod.OFF_PEAK: 0.3
        }
        
        return carbon_rates[period]
    
    def calculate_carbon_footprint(self, kwh: float) -> float:
        """Calculate carbon footprint for consumption"""
        return kwh * self.get_carbon_intensity()

File: ai_engine/backend/backend/src/main/java/com/ecosync/service/PricingService.java
package com.ecosync.service;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import java.time.DayOfWeek;
import java.time.LocalDate;
import java.time.LocalTime;
import java.time.YearMonth;
import java.util.*;

/**
 * Utility Pricing Service for electricity cost optimization
 * Supports time-of-use pricing and real-time pricing integration
 */
@Service
public class PricingService {

    @Value("${utility.price.per.kwh:0.12}")
    private double basePricePerKwh;

    @Value("${utility.currency:USD}")
    private String currency;

    // Time-of-use pricing tiers (price per kWh)
    private final Map<TimePeriod, Double> timeOfUsePrices;
    
    // Cached daily prices
    private final Map<LocalDate, Map<Integer, Double>> dailyPrices;

    public PricingService() {
        this.timeOfUsePrices = new EnumMap<>(TimePeriod.class);
        this.dailyPrices = new HashMap<>();
        
        // Initialize default time-of-use prices
        initializeDefaultPrices();
    }

    private enum TimePeriod {
        PEAK,      // Highest price
        MID_PEAK,  // Medium price
        OFF_PEAK   // Lowest price
    }

    private void initializeDefaultPrices() {
        // Default TOU pricing (California-style)
        timeOfUsePrices.put(TimePeriod.PEAK, basePricePerKwh * 1.5);    // $0.18
        timeOfUsePrices.put(TimePeriod.MID_PEAK, basePricePerKwh * 1.2); // $0.144
        timeOfUsePrices.put(TimePeriod.OFF_PEAK, basePricePerKwh * 0.7); // $0.084
    }

    /**
     * Get current price per kWh based on time-of-use
     */
    public double getCurrentPricePerKwh() {
        LocalTime now = LocalTime.now();
        DayOfWeek dayOfWeek = LocalDate.now().getDayOfWeek();
        
        return getPriceForTime(now, dayOfWeek);
    }

    /**
     * Get price for a specific time
     */
    public double getPriceForTime(LocalTime time, DayOfWeek dayOfWeek) {
        TimePeriod period = getTimePeriod(time, dayOfWeek);
        return timeOfUsePrices.get(period);
    }

    /**
     * Determine time period based on time and day
     */
    private TimePeriod getTimePeriod(LocalTime time, DayOfWeek dayOfWeek) {
        boolean isWeekend = dayOfWeek == DayOfWeek.SATURDAY || dayOfWeek == DayOfWeek.SUNDAY;
        
        if (isWeekend) {
            // Weekend: mostly off-peak
            if (time.isBefore(LocalTime.of(6, 0)) || time.isAfter(LocalTime.of(22, 0))) {
                return TimePeriod.OFF_PEAK;
            }
            return TimePeriod.MID_PEAK;
        }
        
        // Weekday pricing
        if (time.isBefore(LocalTime.of(6, 0)) || time.isAfter(LocalTime.of(21, 0))) {
            return TimePeriod.OFF_PEAK;
        }
        
        if (time.isBefore(LocalTime.of(10, 0)) || time.isAfter(LocalTime.of(17, 0))) {
            return TimePeriod.MID_PEAK;
        }
        
        // Peak hours: 10 AM - 5 PM on weekdays
        return TimePeriod.PEAK;
    }

    /**
     * Calculate cost of current consumption
     */
    public double calculateRunningCost(double wattage, double durationHours) {
        double kwh = wattage / 1000 * durationHours;
        return kwh * getCurrentPricePerKwh();
    }

    /**
     * Calculate daily cost estimate
     */
    public double calculateDailyCost(double averageWattage) {
        double totalCost = 0;
        
        for (int hour = 0; hour < 24; hour++) {
            LocalTime time = LocalTime.of(hour, 0);
            double price = getPriceForTime(time, LocalDate.now().getDayOfWeek());
            double hourlyKwh = averageWattage / 1000;
            totalCost += hourlyKwh * price;
        }
        
        return totalCost;
    }

    /**
     * Calculate monthly cost estimate
     */
    public double calculateMonthlyCost(double averageDailyCost) {
        YearMonth currentMonth = YearMonth.now();
        int daysInMonth = currentMonth.lengthOfMonth();
        return averageDailyCost * daysInMonth;
    }

    /**
     * Get hourly prices for today
     */
    public Map<Integer, Double> getHourlyPricesForToday() {
        LocalDate today = LocalDate.now();
        
        if (dailyPrices.containsKey(today)) {
            return dailyPrices.get(today);
        }
        
        Map<Integer, Double> hourlyPrices = new HashMap<>();
        DayOfWeek dayOfWeek = today.getDayOfWeek();
        
        for (int hour = 0; hour < 24; hour++) {
            LocalTime time = LocalTime.of(hour, 0);
            hourlyPrices.put(hour, getPriceForTime(time, dayOfWeek));
        }
        
        dailyPrices.put(today, hourlyPrices);
        return hourlyPrices;
    }

    /**
     * Get low-cost time windows for scheduling
     */
    public List<TimeWindow> getLowCostWindows() {
        List<TimeWindow> windows = new ArrayList<>();
        Map<Integer, Double> hourlyPrices = getHourlyPricesForToday();
        
        double avgPrice = hourlyPrices.values().stream()
                .mapToDouble(Double::doubleValue)
                .average()
                .orElse(basePricePerKwh);
        
        double lowCostThreshold = avgPrice * 0.9;
        
        TimeWindow currentWindow = null;
        
        for (int hour = 0; hour < 24; hour++) {
            double price = hourlyPrices.get(hour);
            
            if (price <= lowCostThreshold) {
                if (currentWindow == null) {
                    currentWindow = new TimeWindow(hour, hour + 1, price);
                } else {
                    currentWindow.endHour = hour + 1;
                    currentWindow.avgPrice = (currentWindow.avgPrice + price) / 2;
                }
            } else {
                if (currentWindow != null) {
                    windows.add(currentWindow);
                    currentWindow = null;
                }
            }
        }
        
        if (currentWindow != null) {
            windows.add(currentWindow);
        }
        
        return windows;
    }

    /**
     * Get high-cost time windows to avoid
     */
    public List<TimeWindow> getHighCostWindows() {
        List<TimeWindow> windows = new ArrayList<>();
        Map<Integer, Double> hourlyPrices = getHourlyPricesForToday();
        
        double avgPrice = hourlyPrices.values().stream()
                .mapToDouble(Double::doubleValue)
                .average()
                .orElse(basePricePerKwh);
        
        double highCostThreshold = avgPrice * 1.1;
        
        TimeWindow currentWindow = null;
        
        for (int hour = 0; hour < 24; hour++) {
            double price = hourlyPrices.get(hour);
            
            if (price >= highCostThreshold) {
                if (currentWindow == null) {
                    currentWindow = new TimeWindow(hour, hour + 1, price);
                } else {
                    currentWindow.endHour = hour + 1;
                    currentWindow.avgPrice = (currentWindow.avgPrice + price) / 2;
                }
            } else {
                if (currentWindow != null) {
                    windows.add(currentWindow);
                    currentWindow = null;
                }
            }
        }
        
        if (currentWindow != null) {
            windows.add(currentWindow);
        }
        
        return windows;
    }

    /**
     * Get optimal time to run a high-power task
     */
    public TimeWindow getOptimalRunTime(int durationHours) {
        List<TimeWindow> lowCostWindows = getLowCostWindows();
        
        TimeWindow bestWindow = null;
        double bestPrice = Double.MAX_VALUE;
        
        for (TimeWindow window : lowCostWindows) {
            int windowDuration = window.endHour - window.startHour;
            if (windowDuration >= durationHours && window.avgPrice < bestPrice) {
                bestPrice = window.avgPrice;
                bestWindow = window;
            }
        }
        
        return bestWindow;
    }

    /**
     * Calculate potential savings from load shifting
     */
    public Map<String, Object> calculateLoadShiftingSavings(
            double wattage, int durationHours, int fromHour, int toHour) {
        
        LocalDate today = LocalDate.now();
        DayOfWeek dayOfWeek = today.getDayOfWeek();
        
        double fromPrice = getPriceForTime(LocalTime.of(fromHour, 0), dayOfWeek);
        double toPrice = getPriceForTime(LocalTime.of(toHour, 0), dayOfWeek);
        
        double kwh = wattage / 1000 * durationHours;
        double currentCost = kwh * fromPrice;
        double shiftedCost = kwh * toPrice;
        double savings = currentCost - shiftedCost;
        
        Map<String, Object> result = new HashMap<>();
        result.put("current_cost", Math.round(currentCost * 100) / 100.0);
        result.put("shifted_cost", Math.round(shiftedCost * 100) / 100.0);
        result.put("savings", Math.round(savings * 100) / 100.0);
        result.put("savings_pct", fromPrice > 0 ? Math.round(savings / currentCost * 10000) / 100.0 : 0);
        result.put("currency", currency);
        
        return result;
    }

    /**
     * Get pricing summary and recommendations
     */
    public Map<String, Object> getPricingSummary() {
        Map<String, Object> summary = new HashMap<>();
        
        summary.put("current_price_per_kwh", Math.round(getCurrentPricePerKwh() * 1000) / 1000.0);
        summary.put("base_price_per_kwh", basePricePerKwh);
        summary.put("currency", currency);
        summary.put("peak_price", timeOfUsePrices.get(TimePeriod.PEAK));
        summary.put("mid_peak_price", timeOfUsePrices.get(TimePeriod.MID_PEAK));
        summary.put("off_peak_price", timeOfUsePrices.get(TimePeriod.OFF_PEAK));
        
        List<TimeWindow> lowCostWindows = getLowCostWindows();
        List<TimeWindow> highCostWindows = getHighCostWindows();
        
        summary.put("low_cost_windows", lowCostWindows);
        summary.put("high_cost_windows", highCostWindows);
        
        if (!lowCostWindows.isEmpty()) {
            TimeWindow bestWindow = lowCostWindows.get(0);
            summary.put("best_time_to_consume", 
                String.format("%02d:00 - %02d:00", bestWindow.startHour, bestWindow.endHour));
        }
        
        return summary;
    }

    /**
     * Get carbon intensity estimate (kg CO2 per kWh)
     * Higher during peak hours when dirtier plants are used
     */
    public double getCarbonIntensity() {
        TimePeriod currentPeriod = getTimePeriod(LocalTime.now(), LocalDate.now().getDayOfWeek());
        
        switch (currentPeriod) {
            case PEAK:
                return 0.5; // Higher carbon intensity during peak
            case MID_PEAK:
                return 0.4;
            case OFF_PEAK:
                return 0.3; // Lower carbon intensity off-peak
            default:
                return 0.4;
        }
    }

    /**
     * Calculate carbon footprint for consumption
     */
    public double calculateCarbonFootprint(double kwh) {
        return kwh * getCarbonIntensity();
    }

    // Inner class for time windows
    public static class TimeWindow {
        public int startHour;
        public int endHour;
        public double avgPrice;

        public TimeWindow(int startHour, int endHour, double avgPrice) {
            this.startHour = startHour;
            this.endHour = endHour;
            this.avgPrice = avgPrice;
        }

        public Map<String, Object> toMap() {
            Map<String, Object> map = new HashMap<>();
            map.put("start_hour", startHour);
            map.put("end_hour", endHour);
            map.put("avg_price", avgPrice);
            map.put("time_range", String.format("%02d:00 - %02d:00", startHour, endHour));
            return map;
        }

        @Override
        public String toString() {
            return String.format("%02d:00 - %02d:00 @ $%.3f/kWh", startHour, endHour, avgPrice);
        }
    }
}
